lib/IR/DomTreeUpdater.cpp

```diff
@@ -0,0 +1,366 @@
+  if (hasPendDeletedBB() || hasPendingPostDomTreeUpdates()) {
+    if (hasPendDeletedBB()) {
+      applyDomTreeUpdates();
+      forceFlushDeletedBB();
+    }
+    PDT->recalculate(F);
+    PendPDTUpdateIndex = PendUpdates.size();
+    dropOutOfDateUpdates();
+  }
+}
+
+bool DomTreeUpdater::hasPendingUpdates() const {
+  return hasPendingDomTreeUpdates() || hasPendingPostDomTreeUpdates();
+}
+
+bool DomTreeUpdater::hasPendingDomTreeUpdates() const {
+  if (DT == nullptr)
+    return false;
+  return PendUpdates.size() != PendDTUpdateIndex;
+}
+
+bool DomTreeUpdater::hasPendingPostDomTreeUpdates() const {
+  if (PDT == nullptr)
+    return false;
+  return PendUpdates.size() != PendPDTUpdateIndex;
+}
+
+bool DomTreeUpdater::hasPendingDeletedBB(BasicBlock *DelBB) const {
+  if (Ups == UpdateStrategy::Eager || DeletedBBs.empty())
+    return false;
+  return DeletedBBs.count(DelBB) != 0;
+}
+// The DT and PDT require the nodes related to updates
+// are not deleted when update functions are called.
+// So BasicBlock deletions must be pended when the
+// UpdateStrategy is Lazy. When the UpdateStrategy is
+// Eager, the BasicBlock will be deleted immediately.
+void DomTreeUpdater::deleteBB(BasicBlock *DelBB) {
+  validateDeleteBB(DelBB);
+  if (Ups == UpdateStrategy::Lazy)
+    DeletedBBs.insert(DelBB);
+  else
+    DelBB->eraseFromParent();
+}
+
+void DomTreeUpdater::callbackDeleteBB(BasicBlock *DelBB,
+                                      function_ref<void()> Callback) {
+  validateDeleteBB(DelBB);
+  if (Ups == UpdateStrategy::Lazy) {
+    Callbacks.push_back(CallBackOnDeletion(DelBB, Callback));
+    DeletedBBs.insert(DelBB);
+  } else {
+    DelBB->removeFromParent();
+    Callback();
+    delete DelBB;
+  }
+}
+
+void DomTreeUpdater::validateDeleteBB(BasicBlock *DelBB) {
+  assert(DelBB && "Invalid push_back of nullptr DelBB.");
+  assert(pred_empty(DelBB) && "DelBB has one or more predecessors.");
+  // DelBB is unreachable and all its instructions are dead.
+  while (!DelBB->empty()) {
+    Instruction &I = DelBB->back();
+    // Replace used instructions with an arbitrary value (undef).
+    if (!I.use_empty())
+      I.replaceAllUsesWith(llvm::UndefValue::get(I.getType()));
+    DelBB->getInstList().pop_back();
+  }
+  // Make sure DelBB has a valid terminator instruction. As long as DelBB is a
+  // Child of Function F it must contain valid IR.
+  new UnreachableInst(DelBB->getContext(), DelBB);
+}
+
+void DomTreeUpdater::applyUpdates(ArrayRef<DominatorTree::UpdateType> Updates) {
+  if (Ups == UpdateStrategy::Lazy) {
+    SmallVector<DominatorTree::UpdateType, 8> Seen;
+    for (auto U : Updates)
+      // Avoid duplicates to applyLazyUpdate() to save on analysis.
+      if (std::none_of(Seen.begin(), Seen.end(),
+                       [U](DominatorTree::UpdateType S) { return S == U; })) {
+        Seen.push_back(U);
+        applyLazyUpdate(U.getKind(), U.getFrom(), U.getTo());
+      }
+  } else {
```

Comment:
LLVM prefers early returns for readability reason. You can add a return at the end of the if body and remove the else here.